File: Functions/findCheckupSegments.m
function segments = findCheckupSegments(timeWithGaps, voltage, current, timeS, startTime, endTime)
% findCheckupSegments - Find constant-current discharge segments for checkup analysis
%
% This function identifies constant-current discharge segments within a
% specified time range for battery checkup capacity analysis.
%
% Inputs:
%   timeWithGaps - Datetime array with timestamps (may contain NaT)
%   voltage      - Voltage array in Volts
%   current      - Current array in Amperes
%   timeS        - Time array in seconds (may contain NaN values)
%   startTime    - Start datetime for analysis range
%   endTime      - End datetime for analysis range
%
% Outputs:
%   segments     - Cell array where each cell contains index ranges of
%                  constant-current discharge segments

fprintf('\nStarting checkup capacity analysis...\n');
tic;  % Restart timer

% Select data within the specified datetime range
selectedIndices = (timeWithGaps >= startTime) & (timeWithGaps <= endTime);
selectedTime = timeWithGaps(selectedIndices);
selectedVoltage = voltage(selectedIndices);
selectedCurrent = current(selectedIndices);
selectedTimeS = timeS(selectedIndices);

% Identify segments where current is constant (checkup discharge)
% Target current: -1.16 A (58/50), tolerance: ±0.1 A
constantCurrentValue = -58/50;
tolerance = 0.1;
constantCurrentIndices = abs(selectedCurrent - constantCurrentValue) <= tolerance;

% Segment filtering parameters
minSegmentLength = 2000;  % Minimum data points for valid checkup

% Find constant-current segments that meet minimum length requirement
segments = findSegments(constantCurrentIndices, minSegmentLength);
fprintf('Found %d checkup discharge segments. (Elapsed: %.2f s)\n', length(segments), toc);

end

function segments = findSegments(indices, minLength)
    % Find continuous segments where indices are true, with minimum length
    %
    % Inputs:
    %   indices   - Logical array indicating which points meet criteria
    %   minLength - Minimum number of consecutive true values to form a segment
    %
    % Outputs:
    %   segments  - Cell array where each cell contains index ranges
    
    segments = {};
    startIdx = find(indices, 1, 'first');
    
    while ~isempty(startIdx)
        % Find where the segment ends
        endIdx = find(~indices(startIdx:end), 1, 'first') + startIdx - 2;
        if isempty(endIdx)
            endIdx = length(indices);
        end
        
        % Only keep segments that meet minimum length requirement
        if (endIdx - startIdx + 1) >= minLength
            segments{end+1} = startIdx:endIdx;
        end
        
        % Find next segment start
        startIdx = find(indices(endIdx+1:end), 1, 'first') + endIdx;
    end
end

File: Functions/findCheckupSegments.py
"""
findCheckupSegments - Find constant-current discharge segments for checkup analysis

This module identifies constant-current discharge segments within a
specified time range for battery checkup capacity analysis.

Author: Converted from MATLAB to Python
Date: November 25, 2025
"""

import numpy as np
import pandas as pd
import time


def find_checkup_segments(time_with_gaps, voltage, current, time_s, start_time, end_time):
    """
    Find constant-current discharge segments for checkup analysis.
    
    This function identifies constant-current discharge segments within a
    specified time range for battery checkup capacity analysis.
    
    Parameters
    ----------
    time_with_gaps : numpy.ndarray
        Datetime array with timestamps (may contain NaT)
    voltage : numpy.ndarray
        Voltage array in Volts
    current : numpy.ndarray
        Current array in Amperes
    time_s : numpy.ndarray
        Time array in seconds (may contain NaN values)
    start_time : datetime or np.datetime64
        Start datetime for analysis range
    end_time : datetime or np.datetime64
        End datetime for analysis range
    
    Returns
    -------
    segments : list of numpy.ndarray
        List where each element contains index ranges of
        constant-current discharge segments
    """
    
    print('\nStarting checkup capacity analysis...')
    tic = time.time()  # Start timer
    
    # Select data within the specified datetime range
    selected_indices = (time_with_gaps >= start_time) & (time_with_gaps <= end_time)
    selected_time = time_with_gaps[selected_indices]
    selected_voltage = voltage[selected_indices]
    selected_current = current[selected_indices]
    selected_time_s = time_s[selected_indices]
    
    # Identify segments where current is constant (checkup discharge)
    # Target current: -1.16 A (58/50), tolerance: ±0.1 A
    constant_current_value = -58/50
    tolerance = 0.1
    constant_current_indices = np.abs(selected_current - constant_current_value) <= tolerance
    
    # Segment filtering parameters
    min_segment_length = 2000  # Minimum data points for valid checkup
    
    # Find constant-current segments that meet minimum length requirement
    segments = _find_segments(constant_current_indices, min_segment_length)
    
    elapsed = time.time() - tic
    print(f'Found {len(segments)} checkup discharge segments. (Elapsed: {elapsed:.2f} s)')
    
    return segments


def _find_segments(indices, min_length):
    """
    Find continuous segments where indices are True, with minimum length.
    
    Parameters
    ----------
    indices : numpy.ndarray
        Boolean array indicating which points meet criteria
    min_length : int
        Minimum number of consecutive True values to form a segment
    
    Returns
    -------
    segments : list of numpy.ndarray
        List where each element contains index ranges (as arrays)
    """
    
    segments = []
    
    # Find all indices where True values start
    diff_indices = np.diff(np.concatenate([[False], indices, [False]]).astype(int))
    start_indices = np.where(diff_indices == 1)[0]
    end_indices = np.where(diff_indices == -1)[0]
    
    # Filter segments by minimum length
    for start_idx, end_idx in zip(start_indices, end_indices):
        if (end_idx - start_idx) >= min_length:
            segments.append(np.arange(start_idx, end_idx))
    
    return segments


# For compatibility with import statement in main script
__all__ = ['find_checkup_segments']